pkg/controllers/csi/driver/volumes/app/publisher.go

```diff
@@ -22,163 +22,166 @@ import (
-	bindCfg, err := csivolumes.NewBindConfig(ctx, publisher.db, volumeCfg)
-	if err != nil {
-		return nil, err
-	}
-
-	hasTooManyAttempts, err := publisher.hasTooManyMountAttempts(ctx, bindCfg, volumeCfg)
-	if err != nil {
-		return nil, err
-	}
-
-	if hasTooManyAttempts {
+func (pub *Publisher) PublishVolume(ctx context.Context, volumeCfg *csivolumes.VolumeConfig) (*csi.NodePublishVolumeResponse, error) {
+	if pub.hasRetryLimitReached(volumeCfg) {
 		log.Info("reached max mount attempts for pod, attaching dummy volume, monitoring disabled", "pod", volumeCfg.PodName)
 
 		return &csi.NodePublishVolumeResponse{}, nil
 	}
 
-	if !bindCfg.IsArchiveAvailable() {
+	if !pub.isArchiveAvailable(volumeCfg) {
 		return nil, status.Error(
 			codes.Unavailable,
-			"version or digest is not yet set, csi-provisioner hasn't finished setup yet for tenant: "+bindCfg.TenantUUID,
+			"version or digest is not yet set, csi-provisioner hasn't finished setup yet for DynaKube: "+volumeCfg.DynakubeName,
 		)
 	}
 
-	if err := publisher.ensureMountSteps(ctx, bindCfg, volumeCfg); err != nil {
-		return nil, err
+	if err := pub.mountOneAgent(volumeCfg); err != nil {
+		return nil, status.Error(codes.Internal, fmt.Sprintf("failed to mount oneagent volume: %s", err))
 	}
 
-	agentsVersionsMetric.WithLabelValues(bindCfg.MetricVersionLabel()).Inc()
-
 	return &csi.NodePublishVolumeResponse{}, nil
 }
 
-func (publisher *AppVolumePublisher) UnpublishVolume(ctx context.Context, volumeInfo *csivolumes.VolumeInfo) (*csi.NodeUnpublishVolumeResponse, error) {
-	volume, err := publisher.loadVolume(ctx, volumeInfo.VolumeID)
-	if err != nil {
-		log.Info("failed to load volume info", "error", err.Error())
-	}
+// hasRetryLimitReached creates the base dir for a given app mount if it doesn't exist yet, checks the creation timestamp against the threshold
+// if any of the FS calls fail in an unexpected way, then it is considered that the limit was reached.
+func (pub *Publisher) hasRetryLimitReached(volumeCfg *csivolumes.VolumeConfig) bool {
+	appDir := pub.path.AppMountForID(volumeCfg.VolumeID)
 
-	if volume == nil {
-		return &csi.NodeUnpublishVolumeResponse{}, nil
-	}
+	stat, err := pub.fs.Stat(appDir)
+	if errors.Is(err, os.ErrNotExist) {
+		// First run, create folder, to keep track of time
+		err := pub.fs.MkdirAll(appDir, os.ModePerm)
+		if err != nil {
+			log.Error(err, "failed to create base dir for app mount, skipping injection", "dir", appDir)
 
-	log.Info("loaded volume info", "id", volume.VolumeID, "pod name", volume.PodName, "version", volume.Version, "dynakube", volume.TenantUUID)
+			return true
+		}
 
-	if volume.Version == "" {
-		log.Info("requester has a dummy volume, no node-level unmount is needed")
+		return false
+	} else if err != nil {
+		log.Error(err, "unexpected failure in checking filesystem state, skipping injection", "dir", appDir)
```

Comment:
if the FS in a state that we can't create a folder:
- How do I know how many times we failed at creating a folder?
- How much confidence do we have in that the CSI-driver can ever complete its job?

Reply:
> How do I know how many times we failed at creating a folder?

I wanted to suggest checking the creation time of the binary folder, but that may have the exact same problem. Don't have an answer for this question...

> How much confidence do we have in that the CSI-driver can ever complete its job?

Until now we checked if the CSI driver had issues downloading the binaries and could not get set up, because of networking issues or similar. We still allowed some time for this to resolve itself. If the database access failed for any reason, we errored the mount attempt and tried again. Now we just check the filesystem and stop trying on the first attempt.

What if the node file system is not yet ready but will be for later attempts? I just wanted to bring this up/discuss it, but maybe there is no other way than to fail early here...